lib/MusicBrainz/Server/Edit/Relationship/RelatedEntities.pm

```diff
@@ -17,15 +17,25 @@ around _build_related_entities => sub {
         my $model = type_to_model($type);
         my @ids = @{ $direct->{$type} };
         my @entities = values %{ $self->c->model($model)->get_by_any_ids(@ids) };
-        $self->c->model('ArtistCredit')->load(@entities);
-        $direct->{artist} ||= [];
-        push @{ $direct->{artist} }, map { $_->artist_id }
-            map { $_->artist_credit->all_names }
-                @entities;
+        if ($type ~~ qw( recording release release_group )) {
+            $self->c->model('ArtistCredit')->load(@entities);
+            $direct->{artist} ||= [];
+            push @{ $direct->{artist} }, map { $_->artist_id }
+                map { $_->artist_credit->all_names }
+                    @entities;
+        }
+
+        # For works, we want relationship edits to also appear on the recordings and releases they're linked to
+        my @work_ids = map { $_->id } grep { $_->isa('MusicBrainz::Server::Entity::Work') } @entities;
+        my ($recordings, $hits) = $self->c->model('Recording')->find_by_work(@work_ids);
+        my @work_recording_ids = map { $_->id } @$recordings;
+        push @{ $direct->{recording} }, @work_recording_ids;
+
 
         # For recordings, we want relationship edits to also appear on the releases they're on
         my @recording_ids = map { $_->id } grep { $_->isa('MusicBrainz::Server::Entity::Recording') } @entities;
-        my ($releases, $hits) = $self->c->model('Release')->find_by_recording(@recording_ids);
+        push @recording_ids, @work_recording_ids;
+        my ($releases, $hits) = $self->c->model('Release')->find_by_recording(\@recording_ids);
```

Comment:
I'm honestly not sure why this seems to not work unless I use \@ now. Ideas? :)

Reply:
The old code was broken, from what I can see. It used the first recording id in the list, but then used the second id as `$limit`, the third id as `$offset`, and the rest as `%args`. So it only found releases for the first recording and had random numbers for the limit/offset if there were more than one. The new code looks fine (apparently `find_by_recording` already accepts multiple recording ids in an array ref, despite its name).